psycop/common/sequence_models/tasks.py

```diff
@@ -150,3 +153,141 @@ def configure_optimizers(
+
+        self.classification_head = nn.Linear(self.d_model, num_classes)
+
+        self.metrics = self.create_metrics(num_classes)
+
+    @staticmethod
+    def create_metrics(num_classes: int) -> dict[str, Metric]:
+        is_binary = num_classes == 1
+        if is_binary:
+            return {"AUROC": BinaryAUROC()}
+        return {"AUROC (macro)": MulticlassAUROC(num_classes=num_classes)}
+
+    def training_step(self, batch: BatchWithLabels, batch_idx: int) -> torch.Tensor:  # type: ignore # noqa: ARG002
+        output = self.forward(batch[0], batch[1])
+        loss = output["loss"]
+        self.log_step("Training", output)
+        return loss
+
+    def validation_step(self, batch: BatchWithLabels, batch_idx: int) -> torch.Tensor:  # type: ignore  # noqa: ARG002
+        output = self.forward(inputs=batch[0], labels=batch[1])
+        self.log_step("Validation", output)
+        return output["loss"]
+
+    def log_step(
+        self,
+        mode: Literal["Validation", "Training"],
+        output: dict[str, torch.Tensor],
+    ) -> None:
+        """
+        Logs the metrics for the given mode.
+        """
+        for metric_name, metric in output.items():
+            self.log(f"{mode} {metric_name}", metric)
+
+    def forward(  # type: ignore
+        self,
+        inputs: dict[str, torch.Tensor],
+        labels: torch.Tensor,
+    ) -> dict[str, torch.Tensor]:
+        embedded_patients = self.embedding_module(inputs)
+        encoded_patients = self.encoder_module(**embedded_patients)
+
+        # Aggregate the sequence
+        is_padding = embedded_patients["src_key_padding_mask"]
+        aggregated_patients = self.aggregation_module(
+            encoded_patients,
+            attention_mask=~is_padding,
+        )
+
+        # Classification head
+        logits = self.classification_head(aggregated_patients)
+        loss = self.loss(logits, labels)
+
+        metrics = self.calculate_metrics(logits, labels)
+
+        return {"logits": logits, "loss": loss, **metrics}
+
+    def calculate_metrics(
+        self,
+        logits: torch.Tensor,
+        labels: torch.Tensor,
+    ) -> dict[str, torch.Tensor]:
+        """
+        Calculates the metrics for the task.
+        """
+        # Calculate the metrics
+        metrics = {}
+        if self.is_binary:
+            probs = torch.sigmoid(logits)
+        else:
+            probs = torch.softmax(logits, dim=1)
+
+        for metric_name, metric in self.metrics.items():
+            metrics[metric_name] = metric(probs, labels)
+
+        return metrics
+
+    def collate_fn(
+        self,
+        patients_with_labels: list[tuple[Patient, int]],
+    ) -> tuple[dict[str, torch.Tensor], torch.Tensor]:
+        """
+        Takes a list of patients and returns a dictionary of padded sequence ids.
+        """
+        patients, outcomes = list(zip(*patients_with_labels))  # type: ignore
+        patients: list[Patient] = list(patients)
```

Comment:
We probably want to rewrite using PredictionTimes.

Reply:
moved to issue #263